lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp

```diff
@@ -0,0 +1,708 @@
+    {
+        fNewOpValue = _cvtRotXYZToAttr(x, y, z);
+
+        setValue(fNewOpValue);
+        return true;
+    }
+
+private:
+
+    // Convert from UFE RotXYZ rotation to a value for the transform op.
+    CvtRotXYZToAttrFn _cvtRotXYZToAttr;
+};
+
+}
+
+UsdTransform3dMayaXformStack::UsdTransform3dMayaXformStack(
+    const UsdSceneItem::Ptr& item
+)
+    : UsdTransform3dBase(item), _xformable(prim())
+{
+    TF_AXIOM(_xformable);
+
+    // *** FIXME ***  Consider receiving the ordered transform ops as a ctor
+    // argument, as we're already asking for them.
+
+    // *** FIXME ***  We ask for ordered transform ops, but the prim may have
+    // other transform ops that are not in the ordered list.  However, those
+    // transform ops are not contributing to the final local transform.
+    bool resetsXformStack = false;
+    auto xformOps = _xformable.GetOrderedXformOps(&resetsXformStack);
+    for (const auto& op : xformOps) {
+        std::string opName = op.GetOpName();
+        auto ndx = opNameToNdx.at(opName);
+        _orderedOps[ndx] = op;
+    }
+}
+
+/* static */
+UsdTransform3dMayaXformStack::Ptr UsdTransform3dMayaXformStack::create(
+    const UsdSceneItem::Ptr& item
+)
+{
+    return std::make_shared<UsdTransform3dMayaXformStack>(item);
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::translation() const
+{
+    return getVector3d<GfVec3d>(UsdGeomXformOp::GetOpName(
+        UsdGeomXformOp::TypeTranslate));
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::rotation() const
+{
+    if (!hasOp(NdxRotate)) {
+        return Ufe::Vector3d(0, 0, 0);
+    }
+    UsdGeomXformOp r = getOp(NdxRotate);
+    TF_AXIOM(r);
+
+    CvtRotXYZFromAttrFn cvt = getCvtRotXYZFromAttrFn(r.GetOpName());
+    return cvt(getValue(r.GetAttr(), getTime(path())));
+}
+
+Ufe::Vector3d UsdTransform3dMayaXformStack::scale() const
+{
+    if (!hasOp(NdxRotate)) {
+        return Ufe::Vector3d(1, 1, 1);
+    }
+    UsdGeomXformOp s = getOp(NdxScale);
+    TF_AXIOM(s);
+
+    GfVec3f v;
+    s.Get(&v, getTime(path()));
+    return toUfe(v);
+}
+
+Ufe::TranslateUndoableCommand::Ptr UsdTransform3dMayaXformStack::translateCmd(double x, double y, double z)
+{
+    return setVector3dCmd(GfVec3d(x, y, z),
+        UsdGeomXformOp::GetOpName(UsdGeomXformOp::TypeTranslate));
+}
+
+Ufe::RotateUndoableCommand::Ptr UsdTransform3dMayaXformStack::rotateCmd(double x, double y, double z)
+{
+    // If there is no rotate transform op yet, create it.
+    UsdGeomXformOp r;
+    GfVec3f v(x, y, z);
+    CvtRotXYZToAttrFn cvt = toXYZ; // No conversion is default.
+    if (!hasOp(NdxRotate)) {
+        // Use notification guard, otherwise will generate one notification for
+        // the xform op add, and another for the reorder.
+        InTransform3dChange guard(path());
+        r = _xformable.AddRotateXYZOp();
+        if (!TF_VERIFY(r)) {
+            return nullptr;
+        }
+        r.Set(v);
+        setXformOpOrder();
```

Comment:
Whenever we add a transform op we must reorder, as transform ops are added at the end.